clippy_lints/src/unnecessary_string_new.rs

```diff
@@ -0,0 +1,58 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::{match_def_path, paths};
+use if_chain::if_chain;
+use rustc_errors::Applicability;
+use rustc_hir::{Expr, ExprKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::ty;
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// ### What it does
+    ///
+    /// Detects cases of `&String::new()` being passed as an argument to a function expecting `&str`
+    ///
+    /// ### Why is this bad?
+    ///
+    /// `&String::new()` is longer and may be confusing
+    ///
+    /// ### Example
+    /// ```rust
+    /// vec!["1", "2", "3"].join(&String::new());
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// vec!["1", "2", "3"].join("");
+    /// ```
+    #[clippy::version = "1.61.0"]
+    pub UNNECESSARY_STRING_NEW,
+    style,
+    "detects cases of `&String::new()` being passed as an argument to a function expecting `&str`"
+}
+declare_lint_pass!(UnnecessaryStringNew => [UNNECESSARY_STRING_NEW]);
+
+impl<'tcx> LateLintPass<'tcx> for UnnecessaryStringNew {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>) {
+        let applicability = Applicability::MachineApplicable;
+        if_chain! {
+            if let ExprKind::AddrOf(.., inner_expr) = expr.kind;
+            if let ExprKind::Call(fun, _) = inner_expr.kind;
+            if let ExprKind::Path(ref qpath) = fun.kind;
+            if let Some(fn_def_id) = cx.qpath_res(qpath, fun.hir_id).opt_def_id();
+            if match_def_path(cx, fn_def_id, &paths::STRING_NEW); //  ["alloc", "string", "String", "new"]
```

Comment:
I think a quite easy extension that could be added in this PR is to also lint `String::from("")`. See [Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=ec7975456f076477153bcb51eccbaf59)

Reply:
Will take a look at this, although I'm wondering if there's ever a case where you'd want to use `String::from("")` over `String::new()`, that might be it's own lint. WDYT?

Reply:
Added